### PR TITLE
cephadm: Allow users to specify dashboard port and http/https protocol

### PR DESCRIPTION
**Pull request by jmolmo**
Added to new parameters to the bootstrap command in order to allow the user
select the dashboard port and if it ssl is going to be used or not.

Fixes: https://tracker.ceph.com/issues/44628

Signed-off-by: Juan Miguel Olmo Martínez <jolmomar@redhat.com>

-->
## Checklist
- [ x ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by sebastian-philipp** [comment]
for others to be able to review this:
image

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -716,7 +911,7 @@ class FileLock(object):
         #   https://github.com/benediktschmitt/py-filelock/issues/31
         #   https://stackoverflow.com/questions/17708885/flock-removing-locked-file-without-race-condition
         fd = self._lock_file_fd
-        self._lock_file_fd = None
+        self._lock_file_fd = -1
         fcntl.flock(fd, fcntl.LOCK_UN)
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:915]: Do you know, if `flock` behaves differently when called with `-1` instead of `None`?

File: src/cephadm/cephadm [hunk 2]
```diff
@@ -107,6 +107,201 @@ class Ceph(object):
...
+            entrypoint='/usr/bin/ceph',
+            args=cmd,
+            volume_mounts=mounts,
+        ).run(timeout=timeout)
+
+##################################
+
+class DashboardService(object):
+    """Defines the dashboard service when bootstrap a cluster"""
+    def __init__(self, args = None):
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:224]: what do you think of making `args` mandatory?

File: src/cephadm/cephadm [hunk 3]
```diff
@@ -107,6 +107,201 @@ class Ceph(object):
...
+    def __init__(self, admin_key, uid, gid, config, log_dir, timeout=None):
+        # type: (str,int, int, str,str, Optional[int]) -> None
+        self.admin_key = admin_key
+        self.uid = uid
+        self.gid = gid
+        self.config = config
+        self.log_dir = log_dir
+        self.timeout = timeout
+
+    def ceph(self, cmd, extra_mounts={}, timeout=DEFAULT_TIMEOUT):
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:198]: ha! my renaming this to `__call__`, you can keep the usage compatible.

File: src/cephadm/cephadm [hunk 4]
```diff
@@ -4600,7 +4725,15 @@ def _get_parser():
     parser_bootstrap.add_argument(
         '--initial-dashboard-password',
         help='Initial password for the initial dashboard user')
-
+    parser_bootstrap.add_argument(
+        '--dashboard-ssl-disabled',
+        action='store_true',
+        help='Disable secured connections to the dashboard')
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:4731]: What do you think of keeping ssl unconditionally enabled? would that work for you? or do you have a specific use case?

File: src/cephadm/cephadm [hunk 5]
```diff
@@ -2826,6 +2931,7 @@ def command_deploy():
     else:
         logger.info('%s daemon %s ...' % ('Deploy', args.name))
 
+    daemon_ports = []   # type: List[int]
     if daemon_type in Ceph.daemons:
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:2935]: As you removed the firewall config from deploy_daemon, we have to set the firewall somewhere else.

what about the MON ports etc here? Are they missing?

what about the dashboard ports?

File: src/cephadm/cephadm [hunk 6]
```diff
@@ -107,6 +107,201 @@ class Ceph(object):
...
+        return CephContainer(
+            image=args.image,
+            entrypoint='/usr/bin/ceph',
+            args=cmd,
+            volume_mounts=mounts,
+        ).run(timeout=timeout)
+
+##################################
+
+class DashboardService(object):
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:222]: I'd rename this to 

```suggestion
class BootstrapDashboardService(object):
```

cause, this is not something we want to do *after* the initial bootstrap. Cause then, we really want to do this from the mgr module.

File: src/cephadm/cephadm (hunk 1 above)
Comment (jmolmo) [reply src/cephadm/cephadm:915]: In both cases return an error, because it expects a file descriptor which is a small, non negative integer.

File: src/cephadm/cephadm (hunk 2 above)
Comment (jmolmo) [reply src/cephadm/cephadm:224]: It Is ok

File: src/cephadm/cephadm (hunk 3 above)
Comment (jmolmo) [reply src/cephadm/cephadm:198]: Yes... but i like more to see things like:
`cli.ceph(['mgr', 'module', 'enable', 'dashboard'])`
instead of:
`cli(['mgr', 'module', 'enable', 'dashboard'])`

IMHO, It seems more intuitive, and opens the door to add new methods in the Cli class to execute other kind of tools not using the ceph container.

File: src/cephadm/cephadm (hunk 4 above)
Comment (jmolmo) [reply src/cephadm/cephadm:4731]: I just try to keep intact the dashboard functionality. 
When enabling the dashboard, the user can choose to use or not SSL to access dashboard. Why do we need  to remove this possibility? why do we need to restrict final user freedom?

File: src/cephadm/cephadm (hunk 5 above)
Comment (jmolmo) [reply src/cephadm/cephadm:2935]: I have removed firewall logic from deploy_daemon, because  i would like to convert this function in some moment in the future in something generic used to deploy a container in a host, using one container image, a config file, and a set of mounts.
This future function should be called from other more specific functions like "deploy_mon" , which should take care of specific things for a monitor daemon, as for eaxmple, once deployed the container open the right ports calling the generic Firewall object provided in this PR.

Port opening for monitor, manager, monitoring stack, dashboard are not lost. 
If you search `fw.open_ports` you will see the points where this happens. (most times replacing the rigid `update_firewall` function that I have removed)
Sorry, because I got details about this topic  in the manual tests I did and i forgot to include the information in the PR. ( i will update the PR with it)

Thought about "command_deploy":
In the same way that the best place to have the logic to deploy containers is "cephadm standalone", I think that it is also the right place to have all the firewall management stuff.
It would be nice to "work" always with "service specification" objects ( this objects must store all the information we know about one service) . And pass the reference to the spec object when it would be needed ( from cephadm to mgr/cephadm and reverse).
In mgr/cephadm we can check complex things like ports conflict, collocation, etc and validate the service spec object ...and in cephadm standalone we can just deploy the container using the information provided inside the validated service especification object.

File: src/cephadm/cephadm (hunk 6 above)
Comment (jmolmo) [reply src/cephadm/cephadm:222]: Agree... it will be heavily improved in the future sharing "service information"  between mrg/cephadm and cephadm

Comment (sebastian-philipp) [reply src/cephadm/cephadm:222]: Actually, I hope that we can keep most of the logic in `mgr/cephadm`

**Comment by jmolmo** [comment]
### **Details about manual tests executed:**

## Dashboard with default configuration (SSL port 8443)

**Without firewalld enabled:**
```
# cephadm bootstrap --mon-ip $mon_ip  --initial-dashboard-password admin

INFO:cephadm:Enabling the dashboard module...
INFO:cephadm:Generating a dashboard self-signed certificate...
INFO:cephadm:Creating initial admin user...
INFO:cephadm:Configuring firewall for dashboard service...
INFO:cephadm:Ceph Dashboard is now available at:

	     URL: https://cephlab2-node-00:8443/
	    User: admin
	Password: admin
```

**With firewalld enabled:**
```
[root@cephLab2-node-00 ~]# firewall-cmd --list-all
public (active)
...
  services: dhcpv6-client mdns ssh
  ports: 
...

# cephadm bootstrap --mon-ip $mon_ip  --initial-dashboard-password admin
...
INFO:cephadm:firewalld ready
INFO:cephadm:Enabling firewalld service ceph-mon in current zone...
INFO:cephadm:Enabling firewalld port 3300/tcp in current zone...
INFO:cephadm:Enabling firewalld port 6789/tcp in current zone...
...
INFO:cephadm:Creating mgr...
INFO:cephadm:firewalld ready
INFO:cephadm:Enabling firewalld service ceph in current zone...
INFO:cephadm:Enabling firewalld port 9283/tcp in current zone...
...
INFO:cephadm:Configuring firewall for dashboard service...
INFO:cephadm:firewalld ready
INFO:cephadm:Enabling firewalld port 8443/tcp in current zone...
INFO:cephadm:Ceph Dashboard is now available at:

	     URL: https://cephlab2-node-00:8443/
	    User: admin
	Password: admin

# firewall-cmd --list-all
  public (active)
  ...
    services: ceph ceph-mon dhcpv6-client mdns ssh
    ports: 3300/tcp 6789/tcp 9283/tcp 8443/tcp 9093/tcp 9094/tcp 3000/tcp 9100/tcp 9095/tcp
  ...
  
```
--------------------------------------------------------------------------------
## Dashboard using custom port (SSL port 8900)

**Without firewalld enabled:**
```
# cephadm bootstrap --mon-ip $mon_ip  --initial-dashboard-password admin --dashboard-port 8900
...
INFO:cephadm:Configuring the dashboard module...
INFO:cephadm:Enabling the dashboard module...
INFO:cephadm:Generating a dashboard self-signed certificate...
INFO:cephadm:Creating initial admin user...
INFO:cephadm:Configuring firewall for dashboard service...
INFO:cephadm:Ceph Dashboard is now available at:

	     URL: https://cephlab2-node-00:8900/
	    User: admin
	Password: admin

```
**With firewalld enabled:**

```
[root@cephLab2-node-00 ~]# firewall-cmd --list-all
public (active)
...
  services: dhcpv6-client mdns ssh
  ports: 
...

# cephadm bootstrap --mon-ip $mon_ip  --initial-dashboard-password admin --dashboard-port 8900
...
INFO:cephadm:firewalld ready
INFO:cephadm:Enabling firewalld port 8900/tcp in current zone...
INFO:cephadm:Ceph Dashboard is now available at:

	     URL: https://cephlab2-node-00:8900/
	    User: admin
	Password: admin

  [root@cephLab2-node-00 ~]# firewall-cmd --list-all
  public (active)
...
    services: ceph ceph-mon dhcpv6-client mdns ssh
    ports: 3300/tcp 6789/tcp 9283/tcp 9093/tcp 9094/tcp 3000/tcp 9100/tcp 9095/tcp 8900/tcp
...

```
---------------------------------------------------------------------------------

## Dashboard using custom port and not https (port 8901)

**Without firewalld enabled:**
```
# cephadm bootstrap --mon-ip $mon_ip  --initial-dashboard-password admin --dashboard-port 8901 --dashboard-ssl-disabled
...
INFO:cephadm:Configuring the dashboard module...
INFO:cephadm:Enabling the dashboard module...
INFO:cephadm:Generating a dashboard self-signed certificate...
INFO:cephadm:Creating initial admin user...
INFO:cephadm:Configuring firewall for dashboard service...
INFO:cephadm:Ceph Dashboard is now available at:

	     URL: http://cephlab2-node-00:8901/
	    User: admin
	Password: admin

```
**With firewalld enabled:**

```
[root@cephLab2-node-00 ~]# firewall-cmd --list-all
public (active)
...
  services: dhcpv6-client mdns ssh
  ports: 
...

# cephadm bootstrap --mon-ip $mon_ip  --initial-dashboard-password admin --dashboard-port 8901 --dashboard-ssl-disabled

INFO:cephadm:firewalld ready
INFO:cephadm:Enabling firewalld port 8901/tcp in current zone...
INFO:cephadm:Ceph Dashboard is now available at:

	     URL: http://cephlab2-node-00:8901/
	    User: admin
	Password: admin


[root@cephLab2-node-00 ~]# firewall-cmd --list-all
public (active)
  services: ceph ceph-mon dhcpv6-client mdns ssh
  ports: 3300/tcp 6789/tcp 9283/tcp 9093/tcp 9094/tcp 3000/tcp 9100/tcp 9095/tcp 8901/tcp


```

File: src/cephadm/cephadm (hunk 6 above)
Comment (jmolmo) [reply src/cephadm/cephadm]: You are right... with this change it will be more easy to move "dashboard service" to mgr/cephadm. (encapsulation is a great thing!)

File: src/cephadm/cephadm (hunk 5 above)
Comment (sebastian-philipp) [reply src/cephadm/cephadm:2957]: > I have removed firewall logic from deploy_daemon, because i would like to convert this function in some moment in the future in something generic used to deploy a container in a host, using one container image, a config file, and a set of mounts.
> This future function should be called from other more specific functions like "deploy_mon" , which should take care of specific things for a monitor daemon, as for eaxmple, once deployed the container open the right ports calling the generic Firewall object provided in this PR.

sounds good. Just be aware that `command_deploy` is the entrypoint when calling `cephadm deploy`

> 
> Port opening for monitor, manager, monitoring stack, dashboard are not lost.
> If you search `fw.open_ports` you will see the points where this happens. (most times replacing the rigid `update_firewall` function that I have removed)

Right. saw that.

* you open the MON ports in bootstrap
* you open the MGR ports in bootstrap
* you open the monitoring and NFS ports in `command_deploy`
* you open the monitoring ports in `command_adopt_{prometheus,grafana,alertmanager}`

what I'm missing is:

* Where do you open the ports for other Ceph daemons in `command_deploy`?


> Sorry, because I got details about this topic in the manual tests I did and i forgot to include the information in the PR. ( i will update the PR with it)

Saw that, thanks!

> Thought about "command_deploy":
> In the same way that the best place to have the logic to deploy containers is "cephadm standalone", I think that it is also the right place to have all the firewall management stuff.
> It would be nice to "work" always with "service specification" objects ( this objects must store all the information we know about one service) . And pass the reference to the spec object when it would be needed ( from cephadm to mgr/cephadm and reverse).
> In mgr/cephadm we can check complex things like ports conflict, collocation, etc and validate the service spec object ...and in cephadm standalone we can just deploy the container using the information provided inside the validated service especification object.

sounds great! Just be aware that we have some uncanny compatibility issues between bin/cephadm and mgr/cephadm:

* tasks/cephadm.py in master needs to be compatible to src/cephadm in v15.2.0
* src/cephadm in the octopus branch (and also in master!) needs to be compatible to the latest octopus release

I don't want to start with the compatibility issues of the  `packaged` mode, as this will be really bad. Like:

* Pacific mgr/cephadm will call an octopus bin/cephadm (-2 major versions)

Comment (jmolmo) [reply src/cephadm/cephadm:2957]: It will be acceptable to start the migration to a more friendly system?

we will need to do: 
in src/cephadm:
1. "cephadm deploy" should accept a new parameter  "daemon_spec"  which will include the things we need to run a container. 

- We will start writing in this spec (yaml string) only what we need to manage ports. (firewall service and ports)
In the case of the manager, for example, we can provide only the ports really used ( dashboard + prometheus exporter) , taking into account any kind of user customization.

- We can continue making more rich the spec parameter content, with other properties. The final target will be to create daemons using only this parameter.

in cephadm orchestrator:
2. Add functionality to  "_create_daemon" to deal with the different kind of daemons we can create, in order to generate the right "daemon_spec" content to use in the "cephadm deploy" command.

Comment (sebastian-philipp) [reply src/cephadm/cephadm:2957]: > It will be acceptable to start the migration to a more friendly system?

I guess we have to solve the packaged cephadm mode, before we can start messing with the CLI. 

> 
> we will need to do:
> in src/cephadm:
> 
> 1. "cephadm deploy" should accept a new parameter  "daemon_spec"  which will include the things we need to run a container.
> 
> * We will start writing in this spec (yaml string) only what we need to manage ports. (firewall service and ports)
>   In the case of the manager, for example, we can provide only the ports really used ( dashboard + prometheus exporter) , taking into account any kind of user customization.
> * We can continue making more rich the spec parameter content, with other properties. The final target will be to create daemons using only this parameter.

Note that the [deploy](https://docs.ceph.com/docs/master/man/8/cephadm/#deploy) command CLI is already pretty generic. Maybe we can simply enhance it?

> 
> in cephadm orchestrator:
> 2. Add functionality to "_create_daemon" to deal with the different kind of daemons we can create, in order to generate the right "daemon_spec" content to use in the "cephadm deploy" command.

you mean as in #35471 ?

Comment (jmolmo) [reply src/cephadm/cephadm:2957]: @sebastian-philipp : 
In the "command_deploy" method there is a call at the end to open the ports and create the services with the firewall.

But in any case .... in my view this is wrong because:
- The list of ports/services to open in the firewall should be provided as an specific parameter, or as part of a well defined structure ( the daemon specification) used as parameter in this method . Now we have a generic "config-json" where we can put this information ... but I do not like to use this "bag" to pass critical settings.
- In my view the "command_deploy" should have all the things needed to deploy a container and do not need to do specific actions depending of the type of daemon.
- In the cephadm orchestrator we are using "harcoded" values for ports in several services. To change this implies to create this services using a "service specification", that will generate a "daemon specification" (not as in  #35471, because the things i expect to have in a daemon specification are things like, image, ports, fw_service, mounts, config files, placement ...)
- We will need more changes in cephadm orchestrator services to manage properly service specification provided by users.

This means to generate another huge pull request that probably will be rejected because the amount of code to review, the high risk implied and because it will be very difficult to preview all the collateral effects...

So it seems that the only possibility is to add in the command deploy more specific stuff to deal with the ports of the different kind of daemons.
I think that in this moment it has exactly the same port coverage we had previously. Can you see any gap?

File: src/cephadm/cephadm (hunk 4 above)
Comment (sebastian-philipp) [reply src/cephadm/cephadm:4753]: by being opinionated, we can reduce the complexity of cephadm. we have to think carefully where we want to invest our time in. Users can always disable SSL afterwards, if they want to.

File: src/cephadm/cephadm (hunk 5 above)
Comment (sebastian-philipp) [reply src/cephadm/cephadm:2957]: > I think that in this moment it has exactly the same port coverage we had previously. Can you see any gap?

Yes, Im missing the place where you open the ports for other Ceph daemons in command_deploy. `daemon_ports` is never set within the `if daemon_type in Ceph.daemons` block.

Comment (jmolmo) [reply src/cephadm/cephadm:2957]: > > I think that in this moment it has exactly the same port coverage we had previously. Can you see any gap?
> 
> Yes, Im missing the place where you open the ports for other Ceph daemons in command_deploy. `daemon_ports` is never set within the `if daemon_type in Ceph.daemons` block.

In the case that the daemon_type will be inside the Ceph.daemons, previously we had that the <create_daemon> function calling the "update_firewall" function:

And in this function what we had was:
 - a fw-service added to the firewall if the daemon_type was in (mon, mgr, mds, osd)
 
 We have now exactly the same functionality, but now it is placed in the call to:

` fw.enable_service_for(daemon_type)  At the end of "command_deploy"`

 - A set of ports added if the daemon_type was part of the monitoring_stack daemons or NFSGanesha:

 we have exactly the same functionality in "command_deploy". In fact I didn't touch nothing here. 
 The ports for the monitoring stack daemons or the NFSGanesha daemons are set explicitly into the
 "daemon_ports" vector

 - A set of ports added for the manager.
 Two of this ports were for the dashboard and one for the prometheus node exporter. 
 Cephadm orchestrator does not have the capability to manage the dashboard or the prometheus node exporter, 
 so the ports to open depends if these manager modules (dashboard, prometheus are enabled or not, and the current configuration)
 So the code inside the "update_firewall" function was not correct because it opened always the same ports without taking into account that these ports could be customized.

In order to solve this problem, I have included  the detection of the ports used by manager modules, and passed this list as part of the extra-args param in the cephadm deploy command.

File: src/cephadm/cephadm [hunk 7]
```diff
@@ -112,6 +112,200 @@ class Ceph(object):
...
+
+        logger.info("firewalld ready")
+        self.available = True
+
+    def enable_service_for(self, daemon_type):
+        # type (str) -> None
+        if self.disabled:
+            return
+        if not self.available:
+            logger.debug('Not possible to enable service <%ss>. firewalld.service is not available' % daemon_type)
```
Comment (jschmid1) [inline src/cephadm/cephadm:144]:
```suggestion
            logger.debug('Not possible to enable service <%s>. firewalld.service is not available' % daemon_type)
```

File: src/cephadm/cephadm [hunk 8]
```diff
@@ -112,6 +112,200 @@ class Ceph(object):
...
+
+        fw_services = []
+        if daemon_type == 'mon':
+            fw_services.append('ceph-mon')
+        elif daemon_type in ['mgr', 'mds', 'osd']:
+            fw_services.append('ceph')
+        elif daemon_type == NFSGanesha.daemon_type:
+            fw_services.append('nfs')
+
+        for svc in fw_services:
```
Comment (jschmid1) [inline src/cephadm/cephadm:155]: `enable_service_for` is called with a `type->str` argument, hence `len(fw_services)` should never be `> 1`,  right?

What's the point of using a `list` here?

File: src/cephadm/cephadm [hunk 9]
```diff
@@ -112,6 +112,200 @@ class Ceph(object):
...
+                logger.debug('firewalld port %s is enabled in current zone' % tcp_port)
+
+    def apply_rules(self):
+         # type () -> None
+        if self.disabled or not self.available:
+            return
+
+        call_throws([self.cmd, '--reload'])
+
+class Cli(object):
```
Comment (jschmid1) [inline src/cephadm/cephadm:193]: If a wrapper around `CephContainer` is introduced, we should adapt it in all places. Otherwise we'll end up with in a bad spot where changes to `CephContainer` can break one of the two wrappers.

File: src/cephadm/cephadm [hunk 10]
```diff
@@ -588,7 +782,7 @@ class FileLock(object):
         # os.open() function.
         # This file lock is only NOT None, if the object currently holds the
         # lock.
-        self._lock_file_fd = None
+        self._lock_file_fd = -1 # type: int
```
Comment (jschmid1) [inline src/cephadm/cephadm:785]: Does the changes to `_lock_file_fd` deserve it's own PR?

If not, what's the reason behind this change?

File: src/cephadm/cephadm [hunk 11]
```diff
@@ -112,6 +112,200 @@ class Ceph(object):
 
 ##################################
 
+class Firewalld(object):
```
Comment (jschmid1) [inline src/cephadm/cephadm:115]: Let's make this an abstract class instead. I'm guessing that someone will eventually use a different firewall. 
Using an abstract implementation makes this easier.

Comment (sebastian-philipp) [reply src/cephadm/cephadm:115]: really? sounds a bit overkill to me.

Comment (jmolmo) [reply src/cephadm/cephadm:115]: I think that maybe we should be centered in this moment to in just making it work, and to deal successfully with all the firewall functionality use cases before to start introducing other kind of firewalls.

File: src/cephadm/cephadm (hunk 7 above)
Comment (jmolmo) [reply src/cephadm/cephadm:144]: thx!

File: src/cephadm/cephadm (hunk 8 above)
Comment (jmolmo) [reply src/cephadm/cephadm:155]: You're completely right .. ( this things happens when moving code) thx!

File: src/cephadm/cephadm (hunk 10 above)
Comment (jmolmo) [reply src/cephadm/cephadm:785]: mypy detected the problem.  ( if we use mypy to define types explicitly, it seems sensible to use the right type values when you are using a variable). In this case and can be seen in the documentation a file descriptor must be a small integer number. 
I thought it was ok to solve in this way., also we can tweak mypy annotations in order to solve it ...and allow to merge this pr ( it was failing before)

File: src/cephadm/cephadm (hunk 9 above)
Comment (jmolmo) [reply src/cephadm/cephadm:193]: Do you mean to use always Cli instead CephContainer?

**Comment by sebastian-philipp** [comment]
I've tried to make sense out of this PR by splitting the `Firewall` class into #35901

File: src/cephadm/cephadm (hunk 9 above)
Comment (jschmid1) [reply src/cephadm/cephadm:191]: discussed in the standup

**Comment by jmolmo** [comment]
> I've tried to make sense out of this PR by splitting the `Firewall` class into #35901

What we have in #35901 is exactly the same that we have here but using hardcoded ports for the dashboard. So if we merge #35901 , the next step will be to do exactly the same modifications that we have here to remove the "hardcoded ports". 

I think probably is more straight forward to close #35901 and merge this PR.

**Comment by sebastian-philipp** [comment]
next PR: #36217

**Comment by sebastian-philipp** [comment]
PRs still open

- [ ] `class CLI`
- [ ] `class BootstrapDashboardService`
- [ ] more dynamic setting of ports

**Comment by jmolmo** [comment]
Closed because this PR is replaced by:
https://github.com/ceph/ceph/pull/36716
